src/main/scala/org/scalacheck/Gen.scala

```diff
@@ -449,22 +508,85 @@ object Gen extends GenArities with GenVersionSpecific {
-            BigInt(1, bb.array()) + low
-          }).filter(n => high >= n)
+        chooseBigInteger
+          .choose(low.bigInteger, high.bigInteger)
+          .map(BigInt(_))
+    }
+
+    implicit object chooseBigInteger extends Choose[BigInteger] {
+      def choose(low: BigInteger, high: BigInteger): Gen[BigInteger] =
+        (low compareTo high) match {
+          case n if n > 0 => throw new IllegalBoundsError(low, high)
+          case 0 => Gen.const(low)
+          case _ => /* n < 0 */
+            val span = high.subtract(low).add(BigInteger.ONE)
+            gen((_, seed) => chBigInteger(low, span, seed))
         }
     }
 
+    /**
+     * Choose a BigDecimal number between two given numbers.
+     *
+     * The minimum scale used will be 34. That means that the
+     * fractional part will have at least 34 digits (more if one of
+     * the given numbers has a scale larger than 34).
+     *
+     * The minimum scale was chosen based on Scala's default scale for
+     * expanding infinite fractions:
+     *
+     *   BigDecimal(1) / 3   // 0.3333333333333333333333333333333333
+     *
+     * See chooseBigDecimalScale for more information about scale.
+     */
+    implicit val chooseBigDecimal: Choose[BigDecimal] =
+      chooseBigDecimalScale(minScale = 34)
+
+    /**
+     * The "scale" of a decimal number refers to the number of digits
+     * in the fractional part. For example, 3.0000 has a scale of 4.
+     *
+     * We can generate an arbitrary number of digits in the decimal
+     * expansion of a number, so if a user calls choose(0, 1) we need
+     * to decide "how much" work to do. The minScale ensures that we
+     * do "enough" work to generate interesting numbers.
+     *
+     * The implicit instance fixes this value, but since users may
+     * want to use other scales we expose this method as well.
+     */
+    def chooseBigDecimalScale(minScale: Int): Choose[BigDecimal] =
+      new Choose[BigDecimal] {
+        private val c = chooseJavaBigDecimalScale(minScale)
+        def choose(low: BigDecimal, high: BigDecimal): Gen[BigDecimal] =
+          c.choose(low.bigDecimal, high.bigDecimal).map(BigDecimal(_))
+      }
+
+    /**
+     * Choose a java.math.BigDecimal number between two given numbers.
+     *
+     * See chooseBigDecimal and chooseBigDecimalScale for more comments.
+     */
+    implicit val chooseJavaBigDecimal: Choose[JavaDecimal] =
+      chooseJavaBigDecimalScale(minScale = 34)
+
+    /**
+     * See chooseBigDecimalScale for comments.
+     */
+    def chooseJavaBigDecimalScale(minScale: Int): Choose[JavaDecimal] =
+      new Choose[JavaDecimal] {
+        def choose(low: JavaDecimal, high: JavaDecimal): Gen[JavaDecimal] =
+        (low compareTo high) match {
+          case n if n > 0 => throw new IllegalBoundsError(low, high)
+          case 0 => Gen.const(low)
+          case _ => /* n < 0 */
+            val s = (low.scale max high.scale) max minScale
```

Comment:
It's less explicit, but couldn't the user just specify the minimum desired scale on either of the parameters for `high` and `low` to the `Gen.choose` method?  For example,

```scala
forAll(Gen.choose(BigDecimal(-2).pow(29).setScale(10),
                  BigDecimal( 2).pow(29))) { (d: BigDecimal) =>
  d.scale >= 10
}
```

Reply:
This also came up in the discussion in #637, but I didn't conclude that there's a justification for the `minScale` parameter to the `Choose` factory.

Reply:
You could accomplish this by just changing this line in your pr:

```scala
val s = (low.scale max high.scale) max (low.scale min high.scale)
```

Reply:
I don't understand what that last line accomplishes, the LHS (`low.scale max high.scale`) will always be chosen over the RHS (`low.scale min high.scale`).

My experience here is that many people using `BigDecimal` don't necessarily think about scale. For example, code like 'Gen.choose(BigDecimal(0), BigDecimal(1))` will only produce two possible values, but this issue won't necessarily be detected by authors, leading to more testing.

Using the implicit scale in `BigDecimal` has tripped people up, especially given how tricky it is to reason about. For example, authors might think that `Gen.choose(BigDecimal(0.0000), BigDecimal(1.0000))` will give them a scale of 4, but actually it will be 1 due to a different treatment of `String` and `Double` constructors:

```scala
scala> val x = BigDecimal(0.0000)
val x: scala.math.BigDecimal = 0.0

scala> x.scale
val res0: Int = 1

scala> val y = BigDecimal("0.0000")
val y: scala.math.BigDecimal = 0.0000

scala> y.scale
val res1: Int = 4
```

I'm open to simplifying the API a bit but I strongly feel like we should have a relatively rigorous `minScale`, even if it's hardcoded. It's possible to truncate a generated `BigDecimal` for less precision but the reverse is not possible, so I feel like more precision (to a point) is a better default.